lib/base/logger.hpp

```diff
@@ -104,6 +116,9 @@ class Logger : public ObjectImpl<Logger>
 	static LogSeverity m_ConsoleLogSeverity;
 	static std::mutex m_UpdateMinLogSeverityMutex;
 	static Atomic<LogSeverity> m_MinLogSeverity;
+
+	Atomic<bool> m_CalledOnAllConfigLoaded {false};
+	std::vector<ConfigObject*> m_ObjectFilterCache;
```

Comment:
As these are raw pointer, is there anything preventing them from becoming dangling? (For those in `ConfigObject.m_AllParentsAffectingLogging.Data` one might argue with that they should only be removed in a safe order, but even that should be double checked and here, I don't see anything like that.)

Reply:
Nothing at all and I see no problems as they're only intersected anyway.

Reply:
It looks like the behavior of that is implementation-defined (see https://stackoverflow.com/a/30694084, and even undefined in C++11) and even in the best-case scenario, this could result in some object randomly being logged if is allocated at the same address as some previous object that was included in the filter.

Apart from that, leaving dangling pointers is always dangerous, better hope nobody will ever have an idea that involves dereferencing the pointers in this vector.

Reply:
Seriously, a raw pointer is just a number. Not more, not less. In Log#~Log(🪵) I'm only comparing numbers. Change my mind.

> this could result in some object randomly being logged if is allocated at the same address as some previous object that was included in the filter.

Yes, in case someone who explicitly included an object in a filter removes that object. Again, explicitly. Would you feel better if I maintain an additional vector with Ptr, just for keep alive?

> Apart from that, leaving dangling pointers is always dangerous, better hope nobody will ever have an idea that involves dereferencing the pointers in this vector.

DependencyGraph has the same ~problem~.

Reply:
> Seriously, a raw pointer is just a number. Not more, not less. In Log#~Log(🪵) I'm only comparing numbers. Change my mind.

That's probably what most people expect, but compilers may see an opportunity for optimization that could go wrong.

> > this could result in some object randomly being logged if is allocated at the same address as some previous object that was included in the filter.
> 
> Yes, in case someone who explicitly included an object in a filter removes that object. Again, explicitly. Would you feel better if I maintain an additional vector with Ptr, just for keep alive?

Which would then just prevent the object from ever being freed?

> > Apart from that, leaving dangling pointers is always dangerous, better hope nobody will ever have an idea that involves dereferencing the pointers in this vector.
> 
> DependencyGraph has the same ~problem~.

I hope it doesn't. `DependencyGraph` will even dereference the pointers itself in `DependencyGraph::GetParents()` so there better be no dangling pointers. Objects should do `DependencyGraph::RemoveDependency()` before they die, if they don't, it's a bug.

Reply:
> > Seriously, a raw pointer is just a number. Not more, not less. In Log#~Log(🪵) I'm only comparing numbers. Change my mind.
> 
> That's probably what most people expect, but compilers may see an opportunity for optimization that could go wrong.

Suggestion:

I assert std::is_trivial_v on the pointer type I use.

> > > this could result in some object randomly being logged if is allocated at the same address as some previous object that was included in the filter.
> > 
> > 
> > Yes, in case someone who explicitly included an object in a filter removes that object. Again, explicitly. Would you feel better if I maintain an additional vector with Ptr, just for keep alive?
> 
> Which would then just prevent the object from ever being freed?

Exactly. (Especially with my recursive relationship feature) one shouldn't list every single object in the config.

> > > Apart from that, leaving dangling pointers is always dangerous, better hope nobody will ever have an idea that involves dereferencing the pointers in this vector.
> > 
> > 
> > DependencyGraph has the same ~problem~.
> 
> I hope it doesn't. `DependencyGraph` will even dereference the pointers itself in `DependencyGraph::GetParents()` so there better be no dangling pointers. Objects should do `DependencyGraph::RemoveDependency()` before they die, if they don't, it's a bug.

+1.

All in all, have I ever mentioned I'd like the classic std::shared_ptr back? We'd have std::weak_ptr which enables having not dangling, but not leaking pointers.

Reply:
> Suggestion:
> 
> I assert std::is_trivial_v on the pointer type I use.

Ah, nice! Pointer types are scalar types (https://en.cppreference.com/w/cpp/named_req/ScalarType) which are trivial types (https://en.cppreference.com/w/cpp/named_req/TrivialType).

> compilers may see an opportunity for optimization that could go wrong.

Shall I make my just-numbers `volatile` then?